src/ros2cs/ros2cs_core/Context.cs

```diff
@@ -0,0 +1,187 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+
+namespace ROS2
+{
+    /// <summary>
+    /// ROS Context encapsulating the non-global state of an init/shutdown cycle.
+    /// </summary>
+    /// <remarks>
+    /// If the instance is not disposed it will be shut down by the garbage collector.
+    /// Since the collection tracking the nodes might be finalized at this point
+    /// the handle will be leaked.
+    /// </remarks>
+    public sealed class Context : IContext
+    {
+        /// <remarks>
+        /// Will be disposed on disposal of this instance.
+        /// Furthermore, access to the collection is thread safe.
+        /// </remarks>
+        public IReadOnlyDictionary<string, INode> Nodes { get; private set; }
+
+        /// <inheritdoc/>
+        public bool IsDisposed { get { return !this.Ok(); } }
+
+        /// <inheritdoc/>
+        public event Action OnShutdown;
+
+        /// <summary>
+        /// Handle to the <c>rcl_context_t</c>
+        /// </summary>
+        internal IntPtr Handle { get; private set; } = IntPtr.Zero;
+
+        /// <summary>
+        /// Collection nodes active in this context.
+        /// </summary>
+        /// <remarks>
+        /// Also used for synchronisation when creating / removing nodes.
+        /// </remarks>
+        private Dictionary<string, Node> ROSNodes = new Dictionary<string, Node>();
+
+        /// <summary>
+        /// Get the current RMW implementation.
+        /// </summary>
+        /// <returns>The current implementation as string.</returns>
+        public static string GetRMWImplementation()
+        {
+            return Utils.PtrToString(NativeRmwInterface.rmw_native_interface_get_implementation_identifier());
+        }
+
+        /// <summary>
+        /// Create a new ROS Context.
+        /// </summary>
+        public Context()
+        {
+            this.Nodes = new MappedValueDictionary<string, Node, INode>(
+                new LockedDictionary<string, Node>(this.ROSNodes),
+                node => node
+            );
+            this.Handle = NativeRclInterface.rclcs_get_zero_initialized_context();
+            int ret = NativeRclInterface.rclcs_init(this.Handle, NativeRcl.rcutils_get_default_allocator());
+            if ((RCLReturnEnum)ret != RCLReturnEnum.RCL_RET_OK)
+            {
+                this.FreeHandles();
+                Utils.CheckReturnEnum(ret);
+            }
+        }
+
+        /// <inheritdoc/>
+        /// <remarks>
+        /// This method is thread safe.
+        /// </remarks>
+        public bool Ok()
+        {
+            return NativeRclInterface.rclcs_context_is_valid(this.Handle);
+        }
+
+        /// <inheritdoc/>
+        /// <remarks>
+        /// This method is thread safe.
+        /// </remarks>
+        public bool TryCreateNode(string name, out INode node)
```

Comment:
The context is responsible for creating nodes since it tracks all nodes associated with it to dispose them when it is disposed.
The [docs](http://docs.ros.org/en/humble/p/rcl/generated/function_node_8h_1a404531a35e68e71f4900ff48544682b5.html#function-rcl-node-init) say that a node with the same name would be shut down but also that node name uniqueness is not yet enforced (???).
To be sure I implemented the check to prevent removing valid nodes from the dictionary and let the user deal with name collisions since disposing a node is not thread safe.

Reply:
I ran some tests with ROS2 Humble and the rcl did not complain while the old node remained valid when directly calling the `Node` constructor multiple times with the same name.